File: backend/src/main/java/com/scm/hub/application/service/InventoryService.java
package com.scm.hub.application.service;

import com.scm.hub.domain.model.Product;
import com.scm.hub.domain.model.Stock;
import com.scm.hub.domain.model.Warehouse;
import com.scm.hub.domain.port.InventoryPort;
import com.scm.hub.infrastructure.adapter.rest.WebSocketController;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

/**
 * Service class for managing inventory operations.
 * Acts as an orchestrator between the API layer and the domain ports.
 * This service implements the business logic for managing products, warehouses, and their respective stock levels.
 */
@Service
@RequiredArgsConstructor
public class InventoryService {

    private final InventoryPort inventoryPort;
    private final WebSocketController webSocketController;

    /**
     * Creates a new product in the system.
     * Behavior: Passes the product domain model to the inventory port for persistence.
     * @param product The product details to create. Must not be null.
     * @return The created product with its unique identifier.
     * @throws RuntimeException if SKU is duplicated.
     */
    public Product createProduct(Product product) {
        return inventoryPort.saveProduct(product);
    }

    /**
     * Retrieves all available products in the system.
     * @return A list of products. Returns an empty list if no products are found.
     */
    public List<Product> getAllProducts() {
        return inventoryPort.findAllProducts();
    }

    /**
     * Creates a new warehouse.
     * Behavior: Delegates warehouse creation to the persistence port.
     * @param warehouse The warehouse details. Must not be null.
     * @return The created warehouse.
     */
    public Warehouse createWarehouse(Warehouse warehouse) {
        return inventoryPort.saveWarehouse(warehouse);
    }

    /**
     * Retrieves all warehouses.
     * @return A list of warehouses. Returns an empty list if none exist.
     */
    public List<Warehouse> getAllWarehouses() {
        return inventoryPort.findAllWarehouses();
    }

    /**
     * Updates or creates a stock level for a specific product in a warehouse.
     * Behavior: Constructs a new Stock domain object and persists it via the port.
     * Constraints: 
     * <ul>
     *   <li>productId and warehouseId must refer to existing entities.</li>
     *   <li>quantity should be non-negative.</li>
     * </ul>
     * @param productId The ID of the product. Must not be null.
     * @param warehouseId The ID of the warehouse. Must not be null.
     * @param quantity The new stock quantity. Must be zero or positive.
     * @return The updated stock record.
     * @throws IllegalArgumentException if quantity is negative.
     */
    @org.springframework.transaction.annotation.Transactional
    public Stock updateStock(UUID productId, UUID warehouseId, Integer quantity) {
        if (quantity < 0) {
            throw new IllegalArgumentException("Stock quantity cannot be negative");
        }
        inventoryPort.findProductById(productId).orElseThrow(
                () -> new IllegalArgumentException("Product not found: " + productId));
        inventoryPort.findWarehouseById(warehouseId).orElseThrow(
                () -> new IllegalArgumentException("Warehouse not found: " + warehouseId));

        Stock stock = Stock.builder()
                .productId(productId)
                .warehouseId(warehouseId)
                .quantity(quantity)
                .lastUpdated(LocalDateTime.now())
                .build();
        Stock savedStock = inventoryPort.saveStock(stock);
        webSocketController.broadcastStockUpdate(savedStock);
        return savedStock;
    }

    /**
     * Retrieves the stock levels for all products in a specific warehouse.
     * @param warehouseId The ID of the warehouse. Must not be null.
     * @return A list of stock records. Returns an empty list if the warehouse has no stock or doesn't exist.
     */
    public List<Stock> getStockByWarehouse(UUID warehouseId) {
        return inventoryPort.findStockByWarehouse(warehouseId);
    }

    /**
     * Reduces stock quantity for a specific product in a warehouse.
     * Behavior: Retrieves current stock, reduces by the specified amount, and saves the update.
     * @param productId The ID of the product.
     * @param warehouseId The ID of the warehouse.
     * @param quantityToReduce The amount to reduce. Must be positive.
     * @return The updated stock record.
     * @throws IllegalArgumentException if insufficient stock or invalid parameters.
     */
    @org.springframework.transaction.annotation.Transactional
    public Stock reduceStock(UUID productId, UUID warehouseId, Integer quantityToReduce) {
        if (quantityToReduce <= 0) {
            throw new IllegalArgumentException("Quantity to reduce must be positive");
        }
        Stock currentStock = inventoryPort.findStockByProductAndWarehouse(productId, warehouseId)
                .orElseThrow(() -> new IllegalArgumentException("Stock not found for product and warehouse"));
        if (currentStock.getQuantity() < quantityToReduce) {
            throw new IllegalArgumentException("Insufficient stock: available " + currentStock.getQuantity() + ", requested " + quantityToReduce);
        }
        int newQuantity = currentStock.getQuantity() - quantityToReduce;
        return updateStock(productId, warehouseId, newQuantity);
    }

    /**
     * Searches inventory records based on product metadata and warehouse filters.
     * @param query Optional SKU or product name filter.
     * @param warehouseId Optional warehouse filter.
     * @param minQuantity Minimum stock quantity threshold.
     * @return Matching stock records.
     */
    public List<Stock> searchStock(String query, UUID warehouseId, Integer minQuantity) {
        return inventoryPort.searchStock(query, warehouseId, minQuantity);
    }
}

File: frontend/src/app/core/services/inventory.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product, Warehouse, Stock } from '../models/inventory.model';

/**
 * Service for managing inventory-related operations, including products, warehouses, and stock levels.
 */
@Injectable({
  providedIn: 'root'
})
export class InventoryService {
  private apiUrl = '/api/inventory';

  constructor(private http: HttpClient) { }

  /**
   * Retrieves all available products.
   * @returns An Observable of an array of Product objects.
   */
  getProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/products`);
  }

  /**
   * Retrieves all registered warehouses.
   * @returns An Observable of an array of Warehouse objects.
   */
  getWarehouses(): Observable<Warehouse[]> {
    return this.http.get<Warehouse[]>(`${this.apiUrl}/warehouses`);
  }

  /**
   * Retrieves stock levels for a specific warehouse.
   * @param warehouseId The unique identifier of the warehouse.
   * @returns An Observable of an array of Stock objects for the specified warehouse.
   */
  getStockByWarehouse(warehouseId: string): Observable<Stock[]> {
    return this.http.get<Stock[]>(`${this.apiUrl}/stock/warehouse/${warehouseId}`);
  }

  /**
   * Creates a new product in the system.
   * @param product The product object to be created.
   * @returns An Observable of the created Product object.
   */
  createProduct(product: Product): Observable<Product> {
    return this.http.post<Product>(`${this.apiUrl}/products`, product);
  }

  /**
   * Registers a new warehouse in the system.
   * @param warehouse The warehouse object to be created.
   * @returns An Observable of the created Warehouse object.
   */
  createWarehouse(warehouse: Warehouse): Observable<Warehouse> {
    return this.http.post<Warehouse>(`${this.apiUrl}/warehouses`, warehouse);
  }

  /**
   * Updates the stock level for a product at a specific warehouse.
   * @param productId The ID of the product.
   * @param warehouseId The ID of the warehouse.
   * @param quantity The new absolute quantity of stock.
   * @returns An Observable of the updated Stock object.
   */
  updateStock(productId: string, warehouseId: string, quantity: number): Observable<Stock> {
    return this.http.post<Stock>(`${this.apiUrl}/stock`, null, {
      params: { productId, warehouseId, quantity: quantity.toString() }
    });
  }

  /**
   * Searches for stock levels based on a query, warehouse, and minimum quantity.
   * @param query Search term (e.g., product name or SKU).
   * @param warehouseId Optional filter by warehouse.
   * @param minQuantity Optional filter for minimum stock quantity.
   * @returns An Observable of an array containing matched stock information.
   */
  searchStock(query: string, warehouseId?: string, minQuantity?: number): Observable<any[]> {
    let params: any = {};
    if (query) params.query = query;
    if (warehouseId) params.warehouseId = warehouseId;
    if (minQuantity !== undefined) params.minQuantity = minQuantity.toString();
    return this.http.get<any[]>(`${this.apiUrl}/search`, { params });
  }
}
